nutanix/resource_nutanix_ndb_profiles.go

```diff
@@ -0,0 +1,1280 @@
+				out := make([]*era.VersionClusterAssociation, len(clster))
+				for _, v := range clster {
+					val := v.(map[string]interface{})
+
+					if p1, ok1 := val["nx_cluster_id"]; ok1 {
+						out = append(out, &era.VersionClusterAssociation{
+							NxClusterID: utils.StringPtr(p1.(string)),
+						})
+					}
+				}
+				req.VersionClusterAssociation = out
+			}
+		}
+	}
+
+	if db, ok := d.GetOk("database_parameter_profile"); ok {
+		req.Properties = buildDatabaseProfileProperties(db.([]interface{}))
+
+		// setting some defaults values which are generated at runtime
+		req.Topology = utils.StringPtr("ALL")
+		req.Type = utils.StringPtr("Database_Parameter")
+		req.SystemProfile = false
+		req.DBVersion = utils.StringPtr("ALL")
+	}
+
+	if sp, ok := d.GetOk("software_profile"); ok {
+		softwareProfile = true
+		splist := sp.([]interface{})
+
+		for _, v := range splist {
+			val := v.(map[string]interface{})
+
+			if tp, ok := val["topology"]; ok {
+				req.Topology = utils.StringPtr(tp.(string))
+
+				// other details
+				req.Type = utils.StringPtr("Software")
+				req.SystemProfile = false
+				req.DBVersion = utils.StringPtr("ALL")
+			}
+
+			if ps, ok := val["postgres_database"]; ok {
+				req.Properties = expandSoftwareProfileProp(ps.([]interface{}))
+			}
+
+			if ac, ok1 := d.GetOk("available_cluster_ids"); ok1 {
+				st := ac.([]interface{})
+				sublist := make([]*string, len(st))
+
+				for a := range st {
+					sublist[a] = utils.StringPtr(st[a].(string))
+				}
+				req.AvailableClusterIds = sublist
+			}
+		}
+	}
+
+	if softwareProfile {
+		resp, er := conn.Service.CreateSoftwareProfiles(ctx, req)
+		if er != nil {
+			return diag.FromErr(er)
+		}
+
+		// Get Operation ID from response of SoftwareProfile  and poll for the operation to get completed.
+		opID := resp.OperationID
+		if opID == utils.StringPtr("") {
+			return diag.Errorf("error: operation ID is an empty string")
+		}
+		opReq := era.GetOperationRequest{
+			OperationID: utils.StringValue(opID),
+		}
+
+		log.Printf("polling for operation with id: %s\n", *opID)
+
+		// Poll for operation here - Operation GET Call
+		stateConf := &resource.StateChangeConf{
+			Pending: []string{"PENDING"},
+			Target:  []string{"COMPLETED", "FAILED"},
+			Refresh: eraRefresh(ctx, conn, opReq),
+			Timeout: d.Timeout(schema.TimeoutCreate),
+			Delay:   eraDelay,
+		}
+
+		if _, errWaitTask := stateConf.WaitForStateContext(ctx); errWaitTask != nil {
+			return diag.Errorf("error waiting for software profile	 (%s) to create: %s", *resp.EntityID, errWaitTask)
+		}
+		d.SetId(*resp.EntityID)
+	} else {
+		resp, err := conn.Service.CreateProfiles(ctx, req)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		d.SetId(*resp.ID)
+	}
+
+	// Now if published is present args
+
+	if publish, ok := d.GetOk("published"); ok {
+		req := &era.ProfileRequest{}
+		netReq := &era.UpdateProfileRequest{}
+
+		req.Published = publish.(bool)
+		res, err := conn.Service.GetProfiles(ctx, "", "", d.Id(), "")
```

Comment:
This function call is not readable.
Currently GetProfiles takes params: engine string, profileType string, id string and name string  as input.  We can create a 'filters' struct for this method having above params as attributes. Future changes will be easy.

Reply:
Something like GetProfile(filters *profileFilters)

Reply:
Implemented.